#Домашнє завдання до тем apply(), groupby(), pivot_table()

В цьому домашньому завданні продовжуємо працювати з набором даних `supermarket_sales.csv`.

0. Імпортуйте бібліотеку pandas та зчитайте дані у змінну `df` типу `pandas.DataFrame`.

In [3]:
import pandas as pd
df = pd.read_csv('../../Data/mod 3/supermarket_sales.csv')

1. Дослідимо, який філіал супермаркету ('Branch') є найприбутковішим. Для цього знайдіть сумарний прибуток за кожним філіалом і виявіть, який філіал має найвищий.

In [44]:
df['Profit'] = df['gross income'] * df['gross margin percentage'] / 100
df.groupby('Branch')['Profit'].sum().sort_values(ascending = False).round(3)

Branch
C    250.723
A    240.817
B    240.811
Name: Profit, dtype: float64

2. В якому місті знайходиться філіал з найвищим прибутком? Може в тому місці нам розмітисти ще один магазин.  
Знайдіть відповідь за допомогою функціоналу Pandas.

In [183]:
df.groupby('City')['Profit'].sum().sort_values(ascending = False)

City
Naypyitaw    250.722690
Yangon       240.817167
Mandalay     240.811048
Name: Profit, dtype: float64

In [185]:
df.groupby('City')['Profit'].describe()

count      mean       std       min       25%       50%       75%  \
City                                                                           
Mandalay   332.0  0.725334  0.549768  0.042262  0.271964  0.573429  1.055750   
Naypyitaw  328.0  0.764398  0.596737  0.024214  0.275452  0.615464  1.124161   
Yangon     340.0  0.708286  0.525261  0.028786  0.311768  0.546095  1.041589   

                max  
City                 
Mandalay   2.318571  
Naypyitaw  2.364286  
Yangon     2.356667

Considering that the profit amount in this city does not significantly exceed the profitability metrics in other cities, based on these data alone, I would say that it is not advisable to open another store, relying solely on the profit amount. Furthermore, if we check the number of orders in this city, we see that Yangon and Mandalay both exceed Naypyitaw in terms of order volume. Thus, the demand in other cities is not worse.

3.1. Створіть зводну таблицю, яка покаже, скільки покупок (інвойсів) було зроблено в кожній з філій (`Branch`) за різними категоріями товарів. Запишіть таблицю в змінну `invoices_by_category` і виведіть змінну на екран.
Ця таблиця допоможе проаналізувати, в якій філії купують найбільше товарів кожної з категорій.

In [141]:
invoices_by_category = df.pivot_table(index='Branch', columns='Product line', values='Invoice ID', aggfunc='count')
invoices_by_category

Product line  Electronic accessories  Fashion accessories  Food and beverages  \
Branch                                                                          
A                                 60                   51                  58   
B                                 55                   62                  50   
C                                 55                   65                  66   

Product line  Health and beauty  Home and lifestyle  Sports and travel  
Branch                                                                  
A                            47                  65                 59  
B                            53                  50                 62  
C                            52                  45                 45

Очікуваний результат:

![](https://drive.google.com/uc?export=view&id=1rueAdko6S3UxIHGtojetTxlES-EyM6Yb)

3.2. Викристовуючи змінну `invoices_by_category` дайте відповідь програмно (тобто значення треба не просто знайти очима, а вивести за допомогою коду), в якому філіалі магазину (`Branch`) найбільше інвойсів із покупкою товарів категорії "Електронні аксесуари" (`Electronic accessories`)?


In [198]:
invoices_by_category['Electronic accessories'].sort_values(ascending = False)

Branch
A    60
B    55
C    55
Name: Electronic accessories, dtype: int64

4-6. **Творче завдання на розвиток аналітичного мислення**

Крок 1. Сформулюйте ТРИ питання (гіпотези) до наявних даних, які допомогли б вам зрозуміти, які користувачі що, де та коли найбільше/найменше купують, аби дати на основі цих гіпотез рекомендації бізнесу. Звісно питання мають бути не тими, на які ми вже відповіли в завданнях модулю.

Крок 2. Знайдіть відповіді на свої питання з допомогою функціоналу pandas.

Крок 3. Напишіть, як відповідь на це питання може бути використана для прийняття бізнес рішень.   
   
 Питання можуть бути будь-якої складності, але їх має бути 3. Кожне питання оцінюється як 1 завдання. Без виконання цього завдання ДЗ не приймається. Якщо є питання щодо виконання - пишіть у чат 🙌


1. *Which branch has the highest average sales check?* 
It may be worthwhile to focus on sales manager training. If managers in one of the stores are more effective in making customer proposals, selling more products, or achieving higher sales amounts, this could be a key area for improvement.

2. *Which product line generates the most profit in each branch?* 
This information helps identify the most profitable niches. There is a possibility that a significant portion of the budget is spent on purchasing products that are not selling well. Therefore, it might be prudent to consider discontinuing certain product lines.

3. *What is the gender distribution of the primary target audience for each branch?*
Understanding the gender distribution helps in creating targeted promotional offers. For instance, women may be more impulsive in their purchases, while men might prefer reliable, high-quality products with numerous reviews on the site.

1. *Which branch has the highest average sales check?* 

In [203]:
df.groupby(['Branch', 'City'])['Total'].mean().sort_values(ascending = False).round(3)

Branch  City     
C       Naypyitaw    337.100
B       Mandalay     319.873
A       Yangon       312.354
Name: Total, dtype: float64

As observed, Branch C stands out as the leader, effectively engaging with customers and generating an average of 337 units of currency per visitor. In contrast, other branches lag significantly behind. A potential recommendation is to implement customer service training, focusing on skills such as handling objections, increasing sales, and building trust between the salesperson and the customer.

2. *Which product line generates the most profit in each branch?* 

In [243]:
df.pivot_table(index='Branch', columns='Product line', values='Profit', aggfunc='sum')

Product line  Electronic accessories  Fashion accessories  Food and beverages  \
Branch                                                                          
A                          41.535405            37.035167           38.918595   
B                          38.665405            37.218405           34.500881   
C                          43.013548            48.889048           53.893095   

Product line  Health and beauty  Home and lifestyle  Sports and travel  
Branch                                                                  
A                     28.566333           50.832643          43.929024  
B                     45.307619           39.794024          45.324714  
C                     37.676476           31.509190          35.741333

For Branch A, it is evident that the product line 'Health and Beauty' is the least profitable, whereas the profit from 'Home and Lifestyle' products is the highest. It may be worth considering a shift in focus regarding product categories. For Branch C, the least profitable category is 'Home and Lifestyle', while 'Food and Beverages' is the most profitable. It might be advisable to develop the strong categories that generate higher profits. Over time, this could help establish the store’s reputation as having a great selection in, for example, 'Food and Beverages'.

3. *What is the gender distribution of the primary target audience for each branch?*

In [251]:
df.groupby('Branch').Gender.value_counts(normalize=True)

Branch  Gender
A       Male      0.526471
        Female    0.473529
B       Male      0.512048
        Female    0.487952
C       Female    0.542683
        Male      0.457317
Name: Gender, dtype: float64

As we can see, the majority of the audience across all branches are men. At the same time, the proportion of women is also significant. Therefore, it is essential to design promotional offers that target both genders, taking into account their specific preferences and needs.